structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py

```diff
@@ -45,3 +66,31 @@ def forward(
 
         # reduction = "max" (Default)
         return attention_output.amax(dim=3).sum(2)
+
+    def _get_attention_mask(self, n_samples, mask: torch.Tensor):
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
+        batch_size, seq_length = mask.shape
+        attention_mask = torch.BoolTensor(batch_size * n_samples, seq_length, seq_length).fill_(False)
+        attention_mask = attention_mask.to(device=device)
+        masked_length = mask.sum(dim=-1)
+
+        for i in range(batch_size):
+            i_seq_len = masked_length[i]
+            for j in range(i_seq_len):
+                lower_idx, higher_idx = max(0, j - self.M), min(i_seq_len, j + self.M + 1)
+                attention_mask[i:i + n_samples, j, lower_idx:higher_idx] = True
```

Comment:
Don't we need to reshape this somewhere?  
(batch_size * n_samples, seq_length, seq_length) --> (batch_size, n_samples, seq_length, seq_length)

Reply:
Or do we simply change the input shape to be (reverse logic)
(batch_size * n_samples, seq_length, seq_length) from (batch_size, n_samples, seq_length, seq_length)? 
I couldn't find that piece of code yet.

Reply:
Ah, actually found it. You do it when you add `positional embeddings`. 
Separate comment up there (line 50).